homie-controller/src/lib.rs

```diff
@@ -617,3 +616,206 @@ impl From<ParseFloatError> for HandleError {
+        );
+
+        // Discover a new device.
+        assert_eq!(
+            publish(&controller, "base_topic/device_id/$homie", "4.0").await?,
+            Some(Event::DeviceUpdated {
+                device_id: "device_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$fw/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$stats/+", QoS::AtLeastOnce),
+        );
+        assert_eq!(
+            publish(&controller, "base_topic/device_id/$name", "Device name").await?,
+            Some(Event::DeviceUpdated {
+                device_id: "device_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        assert_eq!(
+            publish(&controller, "base_topic/device_id/$state", "ready").await?,
+            Some(Event::DeviceUpdated {
+                device_id: "device_id".to_owned(),
+                has_required_attributes: true
+            })
+        );
+        let mut expected_device = Device::new("device_id", "4.0");
+        expected_device.state = State::Ready;
+        expected_device.name = Some("Device name".to_owned());
+        assert_eq!(
+            controller.devices().get("device_id").unwrap().to_owned(),
+            expected_device
+        );
+
+        // A node on the device.
+        assert_eq!(
+            publish(&controller, "base_topic/device_id/$nodes", "node_id").await?,
+            Some(Event::DeviceUpdated {
+                device_id: "device_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/node_id/+", QoS::AtLeastOnce),
+        );
+        assert_eq!(
+            publish(
+                &controller,
+                "base_topic/device_id/node_id/$name",
+                "Node name"
+            )
+            .await?,
+            Some(Event::NodeUpdated {
+                device_id: "device_id".to_owned(),
+                node_id: "node_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        assert_eq!(
+            publish(
+                &controller,
+                "base_topic/device_id/node_id/$type",
+                "Node type"
+            )
+            .await?,
+            Some(Event::NodeUpdated {
+                device_id: "device_id".to_owned(),
+                node_id: "node_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        let mut expected_node = Node::new("node_id");
```

Comment:
I still don't like these variables that evolve throughout the test. Let me try something.